store/index.go

```diff
@@ -0,0 +1,324 @@
+	Unique    bool
+}
+
+// IndexConfig stores the configuration for a given Index.
+type IndexConfig struct {
+	Path   string
+	Unique bool
+}
+
+// adds an item to the index
+func indexAdd(indexer Indexer, tx ds.Txn, key ds.Key, data []byte) error {
+	indexes := indexer.Indexes()
+	for path, index := range indexes {
+		err := indexUpdate(indexer.BaseKey(), path, index, tx, key, data, false)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// removes an item from the index
+// be sure to pass the data from the old record, not the new one
+func indexDelete(indexer Indexer, tx ds.Txn, key ds.Key, originalData []byte) error {
+	indexes := indexer.Indexes()
+
+	for path, index := range indexes {
+		err := indexUpdate(indexer.BaseKey(), path, index, tx, key, originalData, true)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// adds or removes a specific index on an item
+func indexUpdate(baseKey ds.Key, path string, index Index, tx ds.Txn, key ds.Key, value []byte,
+	delete bool) error {
+
+	valueKey, err := index.IndexFunc(path, value)
+	if valueKey.String() == "" {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	indexKey := indexPrefix.Child(baseKey).ChildString(path).Instance(valueKey.String()[1:])
+
+	data, err := tx.Get(indexKey)
+	if err != nil && err != ds.ErrNotFound {
+		return err
+	}
+
+	indexValue := make(keyList, 0)
+
+	if err != ds.ErrNotFound {
+		if index.Unique && !delete {
+			return ErrUniqueExists
+		}
+	}
+
+	if data != nil {
+		err = DefaultDecode(data, &indexValue)
+		if err != nil {
+			return err
+		}
+	}
+
+	if delete {
+		indexValue.remove(key)
+	} else {
+		indexValue.add(key)
+	}
+
+	if len(indexValue) == 0 {
+		return tx.Delete(indexKey)
+	}
+
+	iVal, err := DefaultEncode(indexValue)
+	if err != nil {
+		return err
+	}
+
+	return tx.Put(indexKey, iVal)
+}
+
+// keyList is a slice of unique, sorted keys([]byte) such as what an index points to
+type keyList [][]byte
+
+func (v *keyList) add(key ds.Key) {
+	b := key.Bytes()
+	i := sort.Search(len(*v), func(i int) bool {
+		return bytes.Compare((*v)[i], b) >= 0
+	})
+
+	if i < len(*v) && bytes.Equal((*v)[i], b) {
+		// already added
+		return
+	}
+
+	*v = append(*v, nil)
+	copy((*v)[i+1:], (*v)[i:])
+	(*v)[i] = b
+}
+
+func (v *keyList) remove(key ds.Key) {
+	b := key.Bytes()
+	i := sort.Search(len(*v), func(i int) bool {
+		return bytes.Compare((*v)[i], b) >= 0
+	})
+
+	if i < len(*v) {
+		copy((*v)[i:], (*v)[i+1:])
+		(*v)[len(*v)-1] = nil
+		*v = (*v)[:len(*v)-1]
+	}
+}
+
+func (v *keyList) in(key ds.Key) bool {
+	b := key.Bytes()
+	i := sort.Search(len(*v), func(i int) bool {
+		return bytes.Compare((*v)[i], b) >= 0
+	})
+
+	return (i < len(*v) && bytes.Equal((*v)[i], b))
+}
+
+// FilterQuery provides a Query Filter function for querying JSON data
+type FilterQuery struct {
+	query *JSONQuery
+}
+
+// Filter is the primary Filter function
+func (f FilterQuery) Filter(e query.Entry) bool {
```

Comment:
Custom Query Filter to do entry filtering within the `ds.Query`. This isn't much of an optimization per se, but it makes the iterator interface easier to use/build.